flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/BenchmarkUtils.java

```diff
@@ -0,0 +1,233 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.benchmark;
+
+import org.apache.flink.api.common.JobExecutionResult;
+import org.apache.flink.ml.api.AlgoOperator;
+import org.apache.flink.ml.api.Estimator;
+import org.apache.flink.ml.api.Model;
+import org.apache.flink.ml.api.Stage;
+import org.apache.flink.ml.benchmark.generator.DataGenerator;
+import org.apache.flink.ml.param.Param;
+import org.apache.flink.ml.param.WithParams;
+import org.apache.flink.ml.util.ParamUtils;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.bridge.java.internal.StreamTableEnvironmentImpl;
+import org.apache.flink.util.InstantiationUtil;
+import org.apache.flink.util.Preconditions;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+/** Utility methods for benchmarks. */
+@SuppressWarnings("unchecked")
+public class BenchmarkUtils {
+    private static final String benchmarkNamePattern = "^[A-Za-z0-9][A-Za-z0-9_\\-]*$";
+
+    /**
+     * Loads benchmark paramMaps from the provided json map.
+     *
+     * @return A map whose key is the names of the loaded benchmarks, value is the parameters of the
+     *     benchmarks.
+     */
+    public static Map<String, Map<String, ?>> parseBenchmarkParams(Map<String, ?> jsonMap) {
+        Preconditions.checkArgument(
+                jsonMap.containsKey("_version") && jsonMap.get("_version").equals(1));
+
+        Map<String, Map<String, ?>> result = new HashMap<>();
+        for (String key : jsonMap.keySet()) {
+            if (!isValidBenchmarkName(key)) {
+                continue;
+            }
+            result.put(key, (Map<String, ?>) jsonMap.get(key));
+        }
+        return result;
+    }
+
+    /**
+     * Checks whether a string is a valid benchmark name.
+     *
+     * <p>A valid benchmark name should only contain English letters, numbers, hyphens (-) and
+     * underscores (_). The name should not start with a hyphen or underscore.
+     */
+    public static boolean isValidBenchmarkName(String name) {
+        return Pattern.matches(benchmarkNamePattern, name);
+    }
+
+    /**
+     * Instantiates a benchmark from its parameter map and executes the benchmark in the provided
+     * environment.
+     *
+     * @return Results of the executed benchmark.
+     */
+    public static BenchmarkResult runBenchmark(
```

Comment:
Would the following implementation be simpler and shorter?

Note that we can change the signature of `instantiateWithParams(...)` to `<T extends WithParams<?>> T instantiateWithParams(Map<String, ?> jsonMap)`

```
    public static BenchmarkResult runBenchmark(
            StreamExecutionEnvironment env, String name, Map<String, ?> params)
            throws Exception {
        Stage<?> stage = instantiateWithParams((Map<String, ?>) params.get("stage"));
        DataGenerator<?> inputsGenerator = instantiateWithParams((Map<String, ?>) params.get("inputs"));
        DataGenerator<?> modelDataGenerator = null;
        if (params.containsKey("modelData")) {
            modelDataGenerator = instantiateWithParams((Map<String, ?>) params.get("modelData"));
        }

        return runBenchmark(env, name, stage, inputsGenerator, modelDataGenerator);
    }

    public static BenchmarkResult runBenchmark(
            StreamExecutionEnvironment env,
            String name,
            Stage<?> stage,
            DataGenerator<?> inputsGenerator,
            DataGenerator<?> modelDataGenerator)
            throws Exception {
        StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);

        Table[] inputTables = inputsGenerator.getData(tEnv);
        if (modelDataGenerator != null) {
            ((Model<?>) stage).setModelData(modelDataGenerator.getData(tEnv));
        }

        Table[] outputTables;
        if (stage instanceof Estimator) {
            outputTables = ((Estimator<?, ?>) stage).fit(inputTables).getModelData();
        } else if (stage instanceof AlgoOperator) {
            outputTables = ((AlgoOperator<?>) stage).transform(inputTables);
        } else {
            throw new IllegalArgumentException("Unsupported Stage class " + stage.getClass());
        }

        for (Table table : outputTables) {
            tEnv.toDataStream(table).addSink(new DiscardingSink<>());
        }

        JobExecutionResult executionResult = env.execute();

        BenchmarkResult result = new BenchmarkResult();
        result.name = name;
        result.executionTimeMillis = (double) executionResult.getNetRuntime(TimeUnit.MILLISECONDS);

        return result;
    }
```